### PR TITLE
Improved OGL and D3D's AA options in UI

### PR DESCRIPTION
**Pull request by aserna3**
This has gone through a significant rewrite since the initial PR. I've modified D3D to not return quality levels. 99% of the time this is just confusing to the user. Nobody needs 30 AA options that are minuscule improvements on the last. Second, I modified the UI to normalize what OGL and D3D display for AA options. I like this implementation because it will scale to more MSAA/SSAA levels if they are added. Third, I gave the SSAA checkbox a proper description. It used to be copy pasted from the MSAA dropdown. "If unsure, select none"? lol. Fourth, both backends now display what AA type is being applied. It was bad UX previously to show MSAA options if the user had SSAA checked. It implied to some that MSAA + ??x SSAA was being applied, which isn't true. Now it shows "SSAA" if it is enabled.

I plan on removing the SSAA checkbox later. I need this merged first so I know what I'm working with to do that though, as this changes some relevant stuff in the backend, and if I fail to properly remove the checkbox, at the very least this is a big UI improvement.

Screenshots for the Dial-up Woman.

On
Off
desc

### CODE REVIEW

**Comment by MaJoRoesch** [comment]
Screenshot please!

**Comment by BhaaLseN** [comment]
I don't really see a reason to have that checkbox there, since its only there to be there? Either put all the options in the list right away, or do something more with the SSAA checkbox (if its not available everywhere, disable it in those cases; and perhaps move it closer to the dropdown since they look rather disconnected at this point).

**Comment by MaJoRoesch** [comment]
The checkbox was there from the previous antialiasing change, and it has to do with how OGL and D3D work. @degasus @HdkR

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 1]
```diff
@@ -368,6 +368,20 @@ VideoConfigDiag::VideoConfigDiag(wxWindow* parent, const std::string &title, con
 	text_aamode = new wxStaticText(page_enh, wxID_ANY, _("Anti-Aliasing:"));
 	choice_aamode = CreateChoice(page_enh, vconfig.iMultisampleMode, wxGetTranslation(aa_desc));
 
+	if (g_video_backend->GetDisplayName() == "OpenGL")
+	{
+		if (vconfig.bSSAA)
+		{
+			const char* caamodes[] = { _trans("None"), "2x SSAA", "4x SSAA", "8x SSAA" };
+			g_Config.backend_info.AAModes.assign(caamodes, caamodes + sizeof(caamodes) / sizeof(*caamodes));
```
Comment (lioncash) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: You can use ArraySize in CommonFuncs instead of the division

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 2]
```diff
@@ -748,3 +766,28 @@ void VideoConfigDiag::PopulatePostProcessingShaders()
...
+	//Check the checkbox status and not the config option because config hasn't changed yet.
+	if (ssaa_checkbox->IsChecked())
+	{
+		const char* caamodes[] = { _trans("None"), "2x SSAA", "4x SSAA", "8x SSAA" };
+		g_Config.backend_info.AAModes.assign(caamodes, caamodes + sizeof(caamodes) / sizeof(*caamodes));
+		vconfig.bSSAA = true;
+	}
+	else
+	{
+		const char* caamodes[] = { _trans("None"), "2x MSAA", "4x MSAA", "8x MSAA" };
```
Comment (lioncash) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: This array should be placed into the class. It's duplicated three times.

**Comment by lioncash** [comment]
Personally, SSAA options should be included in the choices, and they enable when that choice is selected. Having two UI elements for such a thing is kind of 'lol'-level UX.

**Comment by degasus** [comment]
I agree, the SSAA checkbox should be closer to the dropdown, but I don't see what's wrong with a checkbox + an integer value for the samples?

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 3]
```diff
@@ -143,6 +143,9 @@ static wxString stereo_depth_desc = wxTRANSLATE("Controls the separation distanc
 static wxString stereo_convergence_desc = wxTRANSLATE("Controls the distance of the convergence plane. This is the distance at which virtual objects will appear to be in front of the screen.\nA higher value creates stronger out-of-screen effects while a lower value is more comfortable.");
 static wxString stereo_swap_desc = wxTRANSLATE("Swaps the left and right eye. Mostly useful if you want to view side-by-side cross-eyed.\n\nIf unsure, leave this unchecked.");
 
+static constexpr const char* m_msaamodes[4] = { _trans("None"), "2x MSAA", "4x MSAA", "8x MSAA" };
+static constexpr const char* m_ssaamodes[4] = { _trans("None"), "2x SSAA", "4x SSAA", "8x SSAA" };
```
Comment (lioncash) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: these should both have an s_ prefix

**Comment by aserna3** [comment]
I'd like to remove the SSAA checkbox, but the fact that D3D's AA options can be whatever the driver feels like makes actually selecting SSAA and saving that to config hard if not impossible.

It would certainly help if we could make D3D drop the "Quality level" entries. Users shouldn't care about those anyways.

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 4]
```diff
@@ -18,6 +18,7 @@
 #include <wx/slider.h>
 #include <wx/stattext.h>
 
+#include "Common/CommonFuncs.h"
```
Comment (lioncash) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: This isn't necessary anymore.

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 5]
```diff
@@ -748,3 +764,34 @@ void VideoConfigDiag::PopulatePostProcessingShaders()
 	postprocessing_shader.LoadShader(vconfig.sPostProcessingShader);
 	button_config_pp->Enable(postprocessing_shader.HasOptions());
 }
+
+void VideoConfigDiag::OnSSAAClick(wxCommandEvent& event)
+{
+	//Check the checkbox status and not the config option because config hasn't changed yet.
+	if (ssaa_checkbox->IsChecked())
+		vconfig.bSSAA = true;
```
Comment (lioncash) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: `vconfig.bSSAA = ssaa_checkbox->IsChecked();`

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 6]
```diff
@@ -748,3 +764,34 @@ void VideoConfigDiag::PopulatePostProcessingShaders()
 	postprocessing_shader.LoadShader(vconfig.sPostProcessingShader);
 	button_config_pp->Enable(postprocessing_shader.HasOptions());
 }
+
+void VideoConfigDiag::OnSSAAClick(wxCommandEvent& event)
+{
+	//Check the checkbox status and not the config option because config hasn't changed yet.
```
Comment (lioncash) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: space after the //

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 7]
```diff
@@ -370,7 +371,21 @@ VideoConfigDiag::VideoConfigDiag(wxWindow* parent, const std::string &title, con
...
+		{
+			aa_text = mode[0];
+
+			if (vconfig.bSSAA)
+				aa_text += "x SSAA";
+			else
+				aa_text += "x MSAA";
+
+			choice_aamode->AppendString(wxGetTranslation(aa_text));
+		}
```
Comment (lioncash) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: This code is duplicated from the event function.

Comment (aserna3) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: Fixed. Github isn't marking the diff as outdated on my end so~

**Comment by MaJoRoesch** [comment]
This might violate some design guidelines, but is this doable?

ssaa

**Comment by Lightwolf219** [comment]
Would it be possible to use radio buttons instead of a checkbox? I think they would probably make much more sense than a checkbox

**Comment by GameRecover** [comment]
Even two drop down boxes would work.
The first one with the type of AA (SSAA, MSAA), the second one with the multiple (2x, 4x, 16x, etc).

If FXAA is on the table it could be added in the first drop down box in a much later build and when it's selected the second box could be greyed out or something. Not that this is what we're discussing here, I just thought I'd give reason as to why I selected such a method.

**Comment by aserna3** [comment]
Two dropdowns would just be a waste of UI space for a simple toggle.

Checkboxes are used for toggle options.

**Comment by MaJoRoesch** [comment]
I think this is the best option for now, until SSAA can be integrated into the dropdown.

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 8]
```diff
@@ -383,6 +391,7 @@ VideoConfigDiag::VideoConfigDiag(wxWindow* parent, const std::string &title, con
 	const wxString af_choices[] = {"1x", "2x", "4x", "8x", "16x"};
 	szr_enh->Add(new wxStaticText(page_enh, wxID_ANY, _("Anisotropic Filtering:")), 1, wxALIGN_CENTER_VERTICAL, 0);
 	szr_enh->Add(CreateChoice(page_enh, vconfig.iMaxAnisotropy, wxGetTranslation(af_desc), 5, af_choices));
+
```
Comment (degasus) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: unrelated whitespace change

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 9]
```diff
@@ -402,6 +411,7 @@ VideoConfigDiag::VideoConfigDiag(wxWindow* parent, const std::string &title, con
 		szr_pp->Add(choice_ppshader);
 		szr_pp->Add(button_config_pp);
 		szr_enh->Add(szr_pp);
+
```
Comment (degasus) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: unrelated whitespace change

File: Source/Core/VideoBackends/D3D/D3DBase.cpp [hunk 10]
```diff
@@ -182,8 +182,9 @@ std::vector<DXGI_SAMPLE_DESC> EnumAAModes(IDXGIAdapter* adapter)
 			{
 				DXGI_SAMPLE_DESC desc;
 				desc.Count = samples;
-				for (desc.Quality = 0; desc.Quality < quality_levels; ++desc.Quality)
-					_aa_modes.push_back(desc);
+				desc.Quality = 0;
+
+				_aa_modes.push_back(desc);
```
Comment (degasus) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: if (quality_levels > 0)

File: Source/Core/VideoBackends/D3D/D3DBase.cpp [hunk 11]
```diff
@@ -178,13 +178,12 @@ std::vector<DXGI_SAMPLE_DESC> EnumAAModes(IDXGIAdapter* adapter)
...
+			desc.Count = samples;
+
 			if (quality_levels > 0)
-			{
-				DXGI_SAMPLE_DESC desc;
-				desc.Count = samples;
-				for (desc.Quality = 0; desc.Quality < quality_levels; ++desc.Quality)
-					_aa_modes.push_back(desc);
-			}
+				_aa_modes.push_back(desc);
```
Comment (degasus) [inline Source/Core/VideoBackends/D3D/D3DBase.cpp]: Now you've missed "desc.Quality = 0"

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 12]
```diff
@@ -748,3 +752,35 @@ void VideoConfigDiag::PopulatePostProcessingShaders()
...
+	choice_aamode->Clear();
+	for (const std::string& mode : vconfig.backend_info.AAModes)
+	{
+		if (mode.find("None") != std::string::npos)
+		{
+			choice_aamode->AppendString(wxGetTranslation(StrToWxStr(mode)));
+		}
+		else
+		{
+			aa_text = mode[0];
```
Comment (degasus) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: I guess this will fail for 16x MSAA?

File: Source/Core/DolphinWX/VideoConfigDiag.h [hunk 13]
```diff
@@ -278,4 +280,5 @@ class VideoConfigDiag : public wxDialog
 
 	VideoConfig &vconfig;
 	std::string ininame;
+	wxString aa_text;
```
Comment (degasus) [inline Source/Core/DolphinWX/VideoConfigDiag.h]: No need for a global, just use a local variable

File: Source/Core/VideoBackends/D3D/main.cpp [hunk 14]
```diff
@@ -113,7 +113,7 @@ void InitBackendInfo()
 				else if (modes[i].Quality)
```
Comment (BhaaLseN) [inline Source/Core/VideoBackends/D3D/main.cpp]: unless i missed something, ```DXGI_SAMPLE_DESC.Quality``` is always set to 0 at this point, so this branch could technically be removed (see ```DX11::D3D::EnumAAModes```)

File: Source/Core/VideoBackends/D3D/main.cpp [hunk 15]
```diff
@@ -113,7 +113,7 @@ void InitBackendInfo()
...
 					samples = StringFromFormat(_trans("%d samples (quality level %d)"), modes[i].Count, modes[i].Quality);
 				else
-					samples = StringFromFormat(_trans("%d samples"), modes[i].Count);
+					samples = StringFromFormat(_trans("%d"), modes[i].Count);
```
Comment (BhaaLseN) [inline Source/Core/VideoBackends/D3D/main.cpp]: ```Count``` is formatted as number and passed into ```_trans```. Does anyone ever intend to translate a single digit or number? This ```_trans``` call could probably be removed.

File: Source/Core/VideoBackends/OGL/main.cpp [hunk 16]
```diff
@@ -122,7 +122,7 @@ static void InitBackendInfo()
 	g_Config.backend_info.Adapters.clear();
 
 	// aamodes
-	const char* caamodes[] = {_trans("None"), "2x MSAA", "4x MSAA", "8x MSAA"};
+	const char* caamodes[] = {_trans("None"), "2x", "4x", "8x"};
```
Comment (BhaaLseN) [inline Source/Core/VideoBackends/OGL/main.cpp]: Isn't this extra "x" at the end superflous, since ```VideoConfigDiag::RefreshAAList``` adds one?

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 17]
```diff
@@ -748,3 +752,30 @@ void VideoConfigDiag::PopulatePostProcessingShaders()
...
+	vconfig.bSSAA = ssaa_checkbox->IsChecked();
+	RefreshAAList();
+}
+
+void VideoConfigDiag::RefreshAAList()
+{
+	choice_aamode->Clear();
+	const std::string& suffix = vconfig.bSSAA ? "x SSAA" : "x MSAA";
+
+	for (std::string mode : vconfig.backend_info.AAModes)
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: As far as I could see here, ```backend_info.AAModes``` is populated with mere numbers by all the backends; except for the "None" entry.
How about changing that to be a list of samples (or whatever that number stands for), only adding the ones actually being supported (such as 2, 4, 8, 16) and implying "None" (unless there is a backend that does not support "No xSAA" or "0x xSAA"). This would greatly simplify ```DX11::D3D::EnumAAModes```, ```DX11::D3D::InitBackendInfo``` and this method here; along with removing the duplicated "None" translation from all the backends (which would be moved here, in front of the loop, with an explicit translation call rather than an indirect one).

Comment (degasus) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: +1

**Comment by BhaaLseN** [comment]
Does this (yet) change, or account for the fact that "None" is no longer in the list of values inside the backend? IIRC we used to store the _array index_ into the config file, while it is desireable to store the _value_ or _number of samples/counts_ instead. If we're still at the array index part, something else probably has to be changed to make the combobox values and the actual selection inside the backend match up again (ie. a ```-1``` somewhere to correct the index for the missing "None" entry)

**Comment by aserna3** [comment]
That's a bit of a dicey problem because with your request to remove none from the backends and have the index just be offset, this still means that there's no way to turn off AA since the lowest value is 2x. This obviously means that I need to code in something to mean "none". Well, by going with the offsets, -1 could mean none, but that's just adding the thing back in in a less readable fashion (0 == 2x but -1 == none? wut). So, I decided to go half way on this. D3D wanted to return 1x meaning none but I was suppressing that. I stopped blocking that and added 1x to OGL as well for consistency. Yes, this still means that there's a "none" option in the backends, but it has the benefit of being readable as well as allowing the backends to have int arrays of supported options, vs a nasty string.

File: Source/Core/DolphinWX/VideoConfigDiag.cpp [hunk 18]
```diff
@@ -748,3 +752,25 @@ void VideoConfigDiag::PopulatePostProcessingShaders()
...
+
+void VideoConfigDiag::RefreshAAList()
+{
+	choice_aamode->Clear();
+	const std::string& suffix = vconfig.bSSAA ? "x SSAA" : "x MSAA";
+
+	for (int mode : vconfig.backend_info.AAModes)
+	{
+		if (mode == 1)
+			choice_aamode->AppendString("None");
```
Comment (BhaaLseN) [inline Source/Core/DolphinWX/VideoConfigDiag.cpp]: This one isn't translated now. Use ```->AppendString(_("None"))``` instead.

**Comment by degasus** [comment]
LGTM

**Comment by delroth** [comment]
-1 for the design imo. It's really non-obvious what the checkbox does (alter the contents of the list box). Make it two boxes — AA type and AA level?

**Comment by MaJoRoesch** [comment]
Two dropdowns is not very good UX either. The ideal situation is what has already been discussed - both types in the same dropdown. And that's the plan for after this PR.

**Comment by yorkiesandskittles** [comment]
Is there any real reason to disable the SSAA checkbox when AA is set to
None? I think it would make more sense to be able to set your AA type and
THEN set your multiplier, not the other way around.

**Comment by void-ghost** [comment]
As MaJoRoesch explained earlier, it's more like an option for base AA, not a separate mode.

**Comment by yorkiesandskittles** [comment]
I know this, but I was wondering why the checkbox has to be disabled when
AA is set to None. I don't have any true experience in observing how code
works in a given program, but I don't think there should be any reason for
it to be disabled at any given time.

**Comment by void-ghost** [comment]
It's another way to convey to user that this checkbox does nothing if AA is disabled.

**Comment by yorkiesandskittles** [comment]
I'm sure that if they see "None" next to SSAA, they'd get the idea and
change the drop-down option. We don't really have to hold the users' hand.
In fact, disabling the check box in this way would likely have a negative
effect - they'd be wondering why they can't enable SSAA.

**Comment by void-ghost** [comment]
Isn't  that better then them wondering "I set SSAA on, why it still does not working"?

**Comment by MaJoRoesch** [comment]
That is yet _another_ problem with the current system that's solved by having everything in a single dropdown.

None
2xMSAA
4xMSAA
8xMSAA
2xSSAA (or alternatively 2xMSAA+SSAA, we'll see how the bikeshedding goes)
4xSSAA
8xSSAA

That's the plan!

**Comment by void-ghost** [comment]
Oh, I like that idea about 2xMSAA+SSAA form.

**Comment by fallaha56** [comment]
guys, don't want to rain on anyone's parade but D3D AA is (still) broken -there is a scaling issue going on with the XFB that leads to obvious scaling artifacts: looks like staircasing in F-Zero

this had been an issue for many months -i first noticed this when forcing SSAA via nVidiainspector via the sparse grid AA option

**Comment by fallaha56** [comment]
https://bugs.dolphin-emu.org/issues/7942 -this issue
